llama.cpp 2950 (New formula)

Vaibhavs10 opened the pull request:
Adds a formula for compiling llama.cpp from a github release. We expose both the CLI as well as the server through this Formula. The Formula is compliant with the strict brew audit.

<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

chenrui333 commented:
relates to #169894

Vaibhavs10 commented:
Ah! that's my bad! Happy to close this in favor of the other (although I haven't seen much activity on the PR)

What's the best way forward in this case?

SMillerDev commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  test do
+    llama_cli_command = "llama-cli -h"
```
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Vaibhavs10 replied:
Thanks for the advice @SMillerDev - I've added a more comprehensive test in there - such that it loads a model and then checks whether the model was loaded and then further offloaded too. Let me know if that works for you.

julien-c commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,21 @@
+  def install
+    system "make", "DLLAMA_FATAL_WARNINGS=ON", "DLLAMA_METAL_EMBED_LIBRARY=ON", "DLLAMA_CURL=ON"
+
+    bin.install "./main" => "llama-cli"
+    bin.install "./server" => "llama-server"
```
for DX i wish they would merge those two commands into one

Vaibhavs10 replied:
I agree, however, IMO the most use would come from the `server`

Vaibhavs10 commented:
Updated the PR with the following things:

1. Added `autobump` for llama.cpp project (throttling it at 10 checks - since llama.cpp is a pretty active project)
2. Updated the test to be more comprehensive - we now download a model and then check if it was loaded and subsequently offloaded too.

The test ran perfectly for me offline, looking into why it fails on the CI.

Vaibhavs10 commented:
Hi @SMillerDev, All tests pass now. I'm sorry it took ages. I eventually found out that the Runners on GitHub CI don't work as well with Metal.

I think the PR is good to merge if you don't have any other questions! 🚀

SMillerDev commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,41 @@
+      revision: "95fb0aefab568348da159efdd370e064d1b35f97"
+  license "MIT"
+
+  livecheck do
+    throttle 10
+  end
```
I don't think this is needed. We can introduce this later if it's necessary for maintainers

Vaibhavs10 replied:
Thanks for the suggestion!

SMillerDev commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,41 @@
+  depends_on xcode: ["15.0", :build]
+  depends_on arch: :arm64
+  depends_on macos: :ventura
+  depends_on :macos
+  uses_from_macos "curl"
```
```suggestion
  depends_on :macos

  uses_from_macos "curl"
```

Vaibhavs10 commented:
Brilliant! Thanks for the suggestions. I've applied them.
I'm looking forward to getting this merged! 🤗

SMillerDev commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,38 @@
+  def install
+    system "make", "LLAMA_FATAL_WARNINGS=ON", "LLAMA_METAL_EMBED_LIBRARY=ON", "LLAMA_CURL=ON"
+
+    bin.install "./main" => "llama-cli"
+    bin.install "./server" => "llama-server"
```
actually, let's do it like this
```suggestion
    bin.install "main" => "llama-cli"
    bin.install "server" => "llama-server"
```

SMillerDev commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  test do
+    llama_cli_command = ["llama-cli",
+                         "--hf-repo",
+                         "ggml-org/tiny-llamas",
+                         "-m",
+                         "stories15M-q4_0.gguf",
+                         "-n",
+                         "400",
+                         "-p",
+                         "I",
+                         "-ngl",
+                         "0"].join(" ")
+    assert_includes shell_output(llama_cli_command), "<s>"
```
This is a little easier to read:
```suggestion
    llama_cli_command = %w[ llama-cli
                         --hf-repo ggml-org/tiny-llamas
                         -m stories15M-q4_0.gguf
                         -n 400 -p I -ngl 0
    ]
    assert_match "<s>", shell_output(llama_cli_command.join(" "))
```

Vaibhavs10 commented:
Thanks for the suggestions again! 🤗 - I just reasked for a review

carlocab commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,31 @@
+      revision: "95fb0aefab568348da159efdd370e064d1b35f97"
+  license "MIT"
+
+  depends_on xcode: ["15.0", :build]
+  depends_on arch: :arm64
+  depends_on macos: :ventura
```
Where are these requirements documented?

carlocab commented on Formula/l/llama.cpp.rb:
```diff
@@ -0,0 +1,31 @@
+  depends_on xcode: ["15.0", :build]
+  depends_on arch: :arm64
+  depends_on macos: :ventura
+  depends_on :macos
```
```suggestion
```
This doesn't seem correct.
